examples/direct_fidelity_estimation.py

```diff
@@ -133,19 +151,24 @@ def direct_fidelity_estimation(circuit: cirq.Circuit, qubits: List[cirq.Qid],
         cirq.DensityMatrixTrialResult,
         simulator.simulate(circuit)).final_density_matrix
 
-    # TODO(#2639): Sample the Pauli states more efficiently when the circuit
-    # consists of Clifford gates only, as described on page 4 of:
+    # When the circuit consists of Clifford gates only, we can sample the Pauli
+    # states more efficiently as described on page 4 of:
     # https://arxiv.org/abs/1104.4695
-    for P_i in itertools.product([cirq.I, cirq.X, cirq.Y, cirq.Z],
-                                 repeat=n_qubits):
+    if clifford_state is not None:
```

Comment:
instead of taking in `clifford_state`, consider trying to evaluate `circuit` as a clifford circuit and include fall-back logic to the exhaustive method

Reply:
Done.